Moving MATLAB Functions from Simulink to m files

jtgrasb opened the pull request:
I went through and found all MATLAB function blocks within the WEC-Sim blocks and changed them to call an m file of the same name that completed the desired function. I located all of these functions within a new folder called "FunctionsCalledbySimBlocks" within the sources/functions directory. This will allow for easier access to the functions as well as better tracking of their changes.

kmruehl reviewed (changes requested):
@jtgrasb thanks for working on this, this will really help us manage the WEC-Sim library. A few requested revisions:
- save the WEC-Sim library as 2019b, the [current version we support](http://wec-sim.github.io/WEC-Sim/master/user/getting_started.html#matlab-requirements)
- also, when saving the library, save it at the top level, the default is for the library to open up to the level it's saved from

All tests are passing, so the revisions appear to work, but I cannot view the library until it's re-saved. Thanks!

jtgrasb commented:
@kmruehl Thanks for the suggestions! I updated those 2 things and also changed the name of 2 function folders after discussing with Adam. Let me know if you find any other issues.

kmruehl commented:
@jtgrasb it looks like the WEC-Sim library is saved in 2021a. We have a bit of a strange workflow for how to save the library that I hope to fix with sub-libraries. But you'll have to open the library and save it as a prior version, as 2019b. 

![image](https://user-images.githubusercontent.com/4603379/127575702-6b0ff843-bff2-4890-a5d4-51de9feb934e.png)

kmruehl commented on source/functions/SimulinkModelFunctions/irregnLYaw.m:
```diff
@@ -0,0 +1,89 @@
+% yaw at which that last interpolation was performed. If the current
+% relative yaw angle - last interpolated > threshold, the interpolation is
+% performed again. To interpolate every time step, let threshold=0.
+
```
The Simulink function block `IrregnLYaw` has some of this code duplicated, was that intended?

``` matlab

function [Fext,relYawLast,coeffsLastMD,coeffsLastRE,coeffsLastIM] = irregnLYaw(A,w,dofGRD,dirGRD,wGRD,fEHRE,fEHIM, fEHMD, phaseRand,dw,time,WaveDir,WaveSpread, Disp, intThresh, prevYaw, prevCoeffMD, prevCoeffRE, prevCoeffIM)
%#codegen
% Fext is the excitation force output, relYawLast is the angle of relative
% yaw at which that last interpolation was performed. If the current
% relative yaw angle - last interpolated > threshold, the interpolation is
% performed again. To interpolate every time step, let threshold=0.

% caseFlag is for debug, delete when satisfied
A1=bsxfun(@plus,w*time,pi/2);
%initialize outputs
Fext = zeros(1,6);
% Fext = zeros(1,size(fExtRE,3));
relYawLast=0;
coeffsLastMD=zeros(length(w),6);
coeffsLastRE=zeros(length(w),6);
coeffsLastIM=zeros(length(w),6);
% coeffsLastMD=zeros(length(w),size(fExtRE,3));
% coeffsLastRE=zeros(length(w),size(fExtRE,3));
% coeffsLastIM=zeros(length(w),size(fExtRE,3));

[Fext,relYawLast,coeffsLastMD,coeffsLastRE,coeffsLastIM] = irregnLYaw(A,w,dofGRD,dirGRD,wGRD,fEHRE,fEHIM, fEHMD, phaseRand,dw,time,WaveDir,WaveSpread, Disp, intThresh, prevYaw, prevCoeffMD, prevCoeffRE, prevCoeffIM);

end

```

jtgrasb replied:
The irregnLYaw case has to be initialized both within the matlab function from simulink and the m-file to avoid errors. Not exactly sure of the reason for the double initialization, but may be due to the higher dimensions.

kmruehl replied:
@jtgrasb thanks for looking into this. I [added this ](https://github.com/WEC-Sim/WEC-Sim/projects/58#card-66339777 )to the project board as something to debug later.

kmruehl commented on source/functions/SimulinkModelFunctions/regnLYaw.m:
```diff
@@ -0,0 +1,69 @@
+% yaw at which that last interpolation was performed. If the current
+% relative yaw angle - last interpolated > threshold, the interpolation is
+% performed again. To interpolate every time step, let threshold=0. 
+
```
The Simulink block `regNLYaw` also has some parameters initialized, is that intended?

```matlab

function [Fext,relYawLast,coeffsLastMD,coeffsLastRE,coeffsLastIM] = regnLYaw(A,w,dofGRD,dirGRD,fEHRE,fEHIM, fEHMD,time,WaveDir,Disp, intThresh, prevYaw, prevCoeffMD, prevCoeffRE, prevCoeffIM)
%#codegen
% Fext is the excitation force output, relYawLast is the angle of relative
% yaw at which that last interpolation was performed. If the current
% relative yaw angle - last interpolated > threshold, the interpolation is
% performed again. To interpolate every time step, let threshold=0. 

%initialize outputs
Fext = zeros(1,6);
relYawLast=0;
coeffsLastMD=zeros(length(w),6); % should be (1,6) for regular waves
coeffsLastRE=zeros(length(w),6);
coeffsLastIM=zeros(length(w),6);

[Fext,relYawLast,coeffsLastMD,coeffsLastRE,coeffsLastIM] = regnLYaw(A,w,dofGRD,dirGRD,fEHRE,fEHIM, fEHMD,time,WaveDir,Disp, intThresh, prevYaw, prevCoeffMD, prevCoeffRE, prevCoeffIM);

end

```

jtgrasb replied:
After talking with Dominic, it seems the nLYaw cases need to be initialized within the matlab function from Simulink due to the nature of the function. Initialization within the m-file or Simulink led to errors.

kmruehl replied:
Same as above, added to project board

kmruehl commented on source/functions/SimulinkModelFunctions/MoorDyn_matlab_FCN.m:
```diff
@@ -0,0 +1,7 @@
+function FLines  = MoorDyn_matlab_FCN(X,XD,Time,CoupTime)
+
```
This one is also has the outputs initialized prior to calling the function in the Simulink block. 

``` matlab

function FLines  = MoorDyn_matlab_FCN(X,XD,Time,CoupTime)

FLines = zeros(1,6);
FLines  = MoorDyn_matlab_FCN(X,XD,Time,CoupTime);

end

```

kmruehl commented on source/functions/SimulinkModelFunctions/irregExcF.m:
```diff
@@ -0,0 +1,22 @@
+
+A1=bsxfun(@plus,w*time,pi/2);
+%Fext = zeros(1,6);
+Fext = zeros(1,size(fExtRE,3));
```
Please delete commented lines, looks like this was meant for debugging

kmruehl reviewed (commented):
@jtgrasb this is great, thank you! I just had a few minor comments:
- Can we remove some of the output initialization so that we are just calling the function from a script? I've highlighted a few examples of Simulink blocks where the outputs are initialized prior to calling the function. If not, that's alright
- Please remove inline comments meant for debugging. Before those were buried in the Library so we couldn't see them as easily, but now that the functions are tracked as separate `*.m` files we should remove superfluous comments. 

Thank you!

jtgrasb commented:
@kmruehl I've removed all the unnecessary comments and initializations as requested, except for the non-linear yaw cases as explained above.

kmruehl reviewed (approved):
@jtgrasb thank you so much, this is great and all tests passed on my end, so I'll go ahead and merge your PR.
